### PR TITLE
feat(quantization): add segmented adjustment quantization

### LINKED ISSUE #2790: [improve](other): add SAQ quantization support
### Kind of improvement

perf

### Affected area

other

### Summary

VSAG currently provides scalar/product quantizers and RaBitQ, but it does not provide SAQ as an opt-in base quantizer. This proposal adds a native SAQ implementation that combines PCA-based variance ordering, dimension segmentation, per-segment bit allocation, and code adjustment.

The change will:

- support L2, inner product, and cosine;
- integrate SAQ with HGraph, IVF, Pyramid, flatten datacells, bucket datacells, and index persistence;
- expose the average bit budget, segment count, adjustment rounds, PCA, and random-rotation options;
- provide reconstruction, query-to-code and code-to-code distance paths;
- add unit/integration tests, English and Chinese documentation, and reproducible benchmarks.

SAQ will be enabled only when `base_quantization_type` is explicitly set to `saq`. Existing defaults and non-SAQ serialized indexes remain unchanged.

### Expected benefit

At the same complete quantized-record length, SAQ provides more accurate distance estimates than VSAG's production fast RaBitQ baseline.

In the fixed-parameter Release experiment:

- relative distance MSE is reduced by 42.1% on SIFT1M and 96.2% on GIST1M;
- HGraph QPS is 1.91x higher on SIFT1M at Recall@10 >= 0.85;
- HGraph QPS is 2.92x higher on GIST1M at Recall@10 >= 0.91;
- SAQ reaches Recall@10 0.9531 on GIST1M, while RaBitQ reaches 0.9100 at the largest tested search parameter.

The trade-off is higher training and construction cost and a slower raw query-to-code scan. The end-to-end gain comes from reaching a common recall with a smaller `ef_search`, not from making every distance primitive faster.

### Setup

- Release build on Ubuntu 24.04, Intel Xeon Platinum 8375C
- SIFT1M (128 dimensions) and GIST1M (960 dimensions), L2
- 1,000,000 encoded base vectors and 65,536 training vectors
- 4 average bits per dimension
- Equal complete record lengths: 76 bytes/vector on SIFT1M and 492 bytes/vector on GIST1M
- HGraph: `max_degree=32`, `ef_construction=300`, 16 build threads, `use_reorder=false`
- `ef_search=40,80,120,200,400,800,1600`
- Top-10 search with 10,000 timed queries
- SAQ: PCA, automatic segmentation, six adjustment rounds, and deterministic segment rotation
- Baseline: production fast RaBitQ with four base bits and six encoding rounds

### Common-recall results

| Dataset | Method | `ef_search` | Recall@10 | QPS | P50 ms | P99 ms | Serialized index |
| --- | --- | ---: | ---: | ---: | ---: | ---: | ---: |
| SIFT1M | SAQ | 40 | 0.85834 | 87,866.9 | 0.181 | 0.230 | 230,326,378 B |
| SIFT1M | RaBitQ | 120 | 0.85013 | 46,050.2 | 0.353 | 0.481 | 230,305,725 B |
| GIST1M | SAQ | 200 | 0.91700 | 4,427.0 | 3.725 | 4.372 | 650,503,142 B |
| GIST1M | RaBitQ | 1600 | 0.91000 | 1,518.3 | 10.875 | 13.727 | 649,665,521 B |

SAQ relative MSE versus RaBitQ is `4.6466e-4` versus `8.0319e-4` on SIFT1M and `1.7485e-6` versus `4.6437e-5` on GIST1M.

The complete records have identical sizes. Total serialized-index overhead is 0.009% on SIFT1M and 0.129% on GIST1M. Peak build memory increases from 116.22 MB to 324.52 MB on SIFT1M and from 456.28 MB to 716.67 MB on GIST1M.

SAQ construction is 2.36x slower on SIFT1M and 1.69x slower on GIST1M. Its raw four-candidate scan is 18.9% and 55.7% slower, respectively.

At the requested Recall@10 target of 0.95, neither method reaches the target on SIFT1M. On GIST1M, SAQ reaches 0.9531 at `ef_search=400`; RaBitQ reaches only 0.9100 at `ef_search=1600`, so no same-0.95 QPS ratio is reported.

### References

- SAQ paper: https://arxiv.org/abs/2509.12086
- Reference implementation: https://github.com/howarlii/saq
- VSAG quantization documentation: https://vsag.io/docs/en/quantization/index.html
- The proposed PR will include the implementation under `src/quantization/saq_quantization/` and the reproducible evaluation under `benchs/saq/`.

### Additional context

The SAQ paper evaluates an IVF-based, block-transposed FastScan pipeline. VSAG HGraph currently accesses independently addressable datacell records, so this proposal does not claim to reproduce the paper's specialized 32-candidate FastScan layout or its maximum IVF speedups.

### PR DESCRIPTION
**Pull request by Lss-lmj**
## Change Type

- [ ] Bug fix
- [x] New feature
- [ ] Improvement/Refactor
- [x] Documentation
- [ ] CI/Build/Infra

## Linked Issue

- Fixes: #2790

## What Changed

- Add Segmented Code Adjustment Quantization (SAQ) as an opt-in base quantizer for HGraph, IVF,
  Pyramid, flatten datacells and bucket datacells under L2, inner-product and cosine distance.
- Implement correct L2 centering, PCA ordering, aligned dynamic/fixed segmentation, variance-aware
  bit allocation, CAQ code adjustment and deterministic optional rotation (seed `20260825`).
- Support 1--13-bit byte-aligned bit planes and two estimator factors per segment while matching
  RaBitQ's complete-record byte budget.
- Add lookup-based scalar scans, AVX2/AVX512 four-candidate batches, safe L2 threshold termination
  and popcount-based code-to-code moments without inverse transforms in the HGraph build hot path.
- Add persistence configuration/layout validation and tests for numerical behavior, all metrics,
  zero-norm cosine, scans, reconstruction, malformed rotations, quantizer round trips and a real
  HGraph build/search/serialization round trip.
- Correct LAPACKE row-major PCA eigenvector extraction and add a non-diagonal covariance regression.
- Add synchronized English/Chinese documentation and reproducible Release experiments.

Select SAQ with `base_quantization_type: saq`. Optional parameters are `saq_avg_bits`,
`saq_segment_count`, `saq_adjustment_rounds`, `saq_use_pca` and `saq_random_rotation`.

The retained artifacts are in `benchs/saq/results/formal-final-20260825`; the detailed protocol,
full recall curves, limitations and ablations are in `benchs/saq/REPORT.md`.

## Test Evidence

- [ ] `make fmt`
- [ ] `make lint`
- [ ] `make test`
- [x] `make cov`, run tests, and collect coverage
- [x] Other (described below)

Test details:

```text
Release SIFT1M/GIST1M experiment: run_exit_status=0 (all 1,000,000 base vectors)
cmake --build build-release --target saq_quantization_benchmark -j2: passed
saq_quantization_benchmark --help: exit 0
make DEBUG_BUILD_DIR=./build-coverage COMPILE_JOBS=4 cov: passed
coverage unit suite: 85,350,390 assertions in 814 test cases, passed
coverage functional suite: 191 non-HGraph + 95 HGraph test cases, passed
lcov v2.3 full repository: 85.0% lines (33,287/39,151)
lcov v2.3 changed production sources: 88.2% lines (4,097/4,646)
lcov v2.3 changed production diff: 98.6% lines (788/799)
lcov v2.3 SAQ directory: 98.8% lines (742/751), 91.3% functions (94/103)
PR-CI clang-tidy-15 filter: 10/10 changed production .cpp files, zero errors
./build/tests/eval_monitor_test --reporter compact: 33 assertions in 5 test cases, passed
./build/mockimpl/tests_mockimpl --reporter compact: 1 assertion in 1 test case, passed
ASan + UBSan focused suite: 21,100,634 assertions in 18 test cases, passed
TSan focused suite: 21,100,634 assertions in 18 test cases, passed
clang-format-15 --dry-run --Werror <all changed C/C++ files>: passed
git diff --check: passed
bash -n benchs/saq/run.sh benchs/saq/run_ablations.sh: passed
python3 -m py_compile benchs/saq/summarize.py benchs/saq/generate_config.py: passed
```

Coverage was built and collected in the repository's official OpenBLAS CI image. The HGraph
partition was reconciled against all 95 registered non-daily cases and scheduled through completed
Catch2 shards plus exact-name reruns because gcov makes the parameterized Tune/Duplicate cases
exceptionally slow; every case has a successful final execution. Existing recall diagnostics were
warnings, not failures. lcov v2.3 required ignoring a strict consistency diagnostic in the C++
standard library during capture and two existing non-SAQ consistency diagnostics while reading the
filtered trace.

The standalone `scripts/coverage/check_cov.sh` threshold is 90%, so the 85.0% full-repository trace
does not pass that repository-wide check. Against `upstream/main`, however, executable lines in the
changed production diff reach 98.6% coverage, and the SAQ directory reaches 98.8% line coverage.
The published Coverage workflow collects and uploads the trace but does not call the threshold
script. The pull-request CI changed-source clang-tidy-15 command was run instead of the full-source
`make lint` target and passed. TSan used `setarch x86_64 -R` because the host's default address layout
conflicts with the GCC runtime. The `[daily]` suite is outside the pull-request functional-test entry
point and the reported validation scope.

## Compatibility Impact

- API/ABI compatibility: additive configuration and implementation; existing public defaults and
  non-SAQ persistence formats are unchanged.
- Behavior changes: only indexes explicitly configured with `base_quantization_type: saq` use the
  new quantizer; PCA eigenvector extraction is corrected for all callers.

## Performance and Concurrency Impact

- Performance impact: at equal complete-record size, direct relative MSE falls 42.1% on SIFT1M and
  96.2% on GIST1M versus production fast RaBitQ. At common recall, HGraph query throughput improves
  1.91x (SIFT, Recall@10 0.85) and 2.92x (GIST, Recall@10 0.91). Raw batch-4 scans are 18.9% and
  55.7% slower, and index construction is 2.36x and 1.69x slower, respectively. GIST SAQ reaches
  Recall@10 0.9531; neither SIFT method reaches the configured 0.95 target.
- Scan-layout boundary: the paper's FastScan path transposes blocks of 32 candidates, whereas the
  reference single-vector quantizer disables FastScan and VSAG graph datacells require independent
  record access. The implementation therefore does not include the paper's specialized FastScan or
  probabilistic m-bound path; adding it requires a new persisted block layout plus graph
  filtering/reorder integration.
- Concurrency/thread-safety impact: no new shared mutable global state; trained transforms and
  quantizer metadata are read-only during concurrent search.

## Documentation Impact

- [ ] No docs update needed
- [x] Updated docs:
  - [x] `README.md`
  - [ ] `DEVELOPMENT.md`
  - [ ] `CONTRIBUTING.md`
  - [x] Other: `docs/docs/{en,zh}/src/quantization/saq.md`,
    `docs/docs/{en,zh}/src/resources/{index_parameters,saq_quantization_benchmark}.md`,
    `tools/saq_quantization_benchmark/README{,_zh}.md`, `benchs/saq/README.md`,
    `benchs/saq/DESIGN.md`, `benchs/saq/REPORT.md`

## Risk and Rollback

- Risk level: medium. The main risk is the new persisted quantizer state and its configuration
  paths. Structural checks reject inconsistent dimensions, plans, budgets, transform sizes and
  non-finite rotation values; quantizer and HGraph round-trip tests cover valid state. The validation
  scope does not include exhaustive fuzz coverage for arbitrary byte corruption.
- Rollback plan: stop selecting `saq` and revert the feature. Existing quantizers and persisted
  non-SAQ indexes are unaffected.

## Checklist

- [x] I have linked the relevant issue (`#2790`)
- [x] I have added/updated tests for new behavior or bug fixes
- [x] I have considered API compatibility impact
- [x] I have updated docs if behavior/workflow changed
- [x] My commit messages follow project conventions (Conventional Commits, optional `[skip ci]` prefix)

### CODE REVIEW

File: src/quantization/saq_quantization/saq_quantizer.cpp [hunk 1]
```diff
@@ -0,0 +1,1116 @@
...
+                       "invalid serialized SAQ segment plan");
+        covered_dimensions += segment.length;
+    }
+    CHECK_ARGUMENT(
+        covered_dimensions == this->dim_,
+        fmt::format(
+            "serialized SAQ plan covers {} of {} dimensions", covered_dimensions, this->dim_));
+    const uint64_t expected_projection_size = use_pca_ ? this->dim_ * this->dim_ : 0;
+    CHECK_ARGUMENT(projection_matrix_.size() == expected_projection_size,
+                   "serialized SAQ PCA matrix has invalid dimensions");
```
Comment (LHT129) [inline src/quantization/saq_quantization/saq_quantizer.cpp:1109]: [suggestion] The `ComputeDistWithThresholdImpl` function writes the computed distance to `dists[0]` via `ComputeProjectedDistance`, but the return value is `stopped_early` which is always `false` for IP and Cosine metrics (since `enable_threshold` is only set for L2). The function name and signature suggest it returns whether the computation stopped early, but for non-L2 callers the return value is always `false`. Consider adding a comment clarifying that threshold-based early termination is only supported for L2.

```cpp
dists[0] = ComputeProjectedDistance(query_lookup,
                                    query_norms,
                                    query_sums,
                                    codes,
                                    threshold,
                                    metric == MetricType::METRIC_TYPE_L2SQR,  // only L2 enables threshold
                                    &stopped_early);
return stopped_early;
```

**Review by LHT129 — COMMENTED** [review]
## Code Review Summary for PR #2792

This PR introduces **Segmented Code Adjustment Quantization (SAQ)**, a new base quantizer for vector similarity search. It is a substantial contribution (~7900 lines across ~80 files) that is well-structured and follows the project conventions consistently.

### What SAQ brings:
- **Segmented quantization**: Splits vector dimensions into variable-length segments, each encoded with an optimal bit width (1-13 bits) via dynamic programming
- **Code Adjustment Quantization (CAQ)**: Coordinate-level adjustment to minimize reconstruction error using cosine similarity as the optimization objective
- **PCA + random orthogonal rotations**: Per-segment random rotations for energy equalization, with optional PCA preprocessing
- **SIMD-accelerated distance computation**: Batch-four distance computation using AVX2/AVX512 intrinsics and popcount-based code-to-code distance
- **L2 centering**: Mean subtraction for L2 metric to improve quantization quality
- **Threshold-based early termination**: For L2 distance computation in search scenarios

### Strengths:
1. **Clean architecture**: SAQ follows the existing `Quantizer<T>` CRTP pattern, integrates naturally with HGraph, IVF, and Pyramid indexes
2. **Thorough parameter validation**: `CHECK_ARGUMENT` guards at every boundary (constructor, deserialization, training)
3. **Comprehensive test coverage**: Tests cover dynamic/fixed segmentation, CAQ adjustment, all three metrics (L2/IP/Cosine), serialization round-trips, reproducible rotations, code-pair distance, zero-norm cosine, malformed rotations, and parameter validation
4. **Correct PCA fix**: The eigenvector extraction fix (column-major LAPACKE to row-major PCA matrix) and BLAS-accelerated covariance computation are well-done improvements
5. **Deterministic rotations**: Seed-based `RandomOrthogonalTransformer` enables reproducible results
6. **Memory discipline**: Consistent use of `Allocator`-aware `Vector<float>`, no raw `new`/`delete`

### One suggestion:
- **`ComputeDistWithThresholdImpl` semantics**: The function always returns `false` for IP and Cosine metrics since threshold-based early termination is only enabled for L2. A clarifying comment would help future readers understand this design choice.

### Overall:
The code quality is high. The implementation is well-architected, properly validated, and thoroughly tested. This is ready to merge after addressing the minor suggestion above.

File: src/quantization/saq_quantization/saq_quantizer.cpp (hunk 1 above)
Comment (Lss-lmj) [reply src/quantization/saq_quantization/saq_quantizer.cpp:1109]: Thanks. Addressed in 58d628d by documenting that threshold early termination is safe only for L2 because its segment contributions are non-negative; inner product and cosine compute the full distance and return false.

File: src/quantization/saq_quantization/saq_quantizer.cpp [hunk 2]
```diff
@@ -0,0 +1,1114 @@
...
+    }
+    InverseProject(projected.data(), data);
+    return true;
+}
+
+template <MetricType metric>
+float
+SAQQuantizer<metric>::ComputeImpl(const uint8_t* codes1, const uint8_t* codes2) const {
+    if (not this->is_trained_ or codes1 == nullptr or codes2 == nullptr) {
+        throw VsagException(ErrorType::INTERNAL_ERROR, "cannot decode untrained SAQ codes");
```
Comment (LHT129) [inline src/quantization/saq_quantization/saq_quantizer.cpp]: **[suggestion]** `ComputeImpl` throws `VsagException` for null codes / untrained state, while `EncodeOneImpl` (line 664) and `DecodeOneImpl` (line 679) return `false` for the same conditions. Consider making the error handling consistent across all `*Impl` methods — either all throw or all return `false`. The current inconsistency could surprise callers that expect a graceful boolean return from `ComputeImpl`.

File: src/quantization/saq_quantization/saq_quantizer.cpp [hunk 3]
```diff
@@ -0,0 +1,1114 @@
...
+    uint8_t bits{0};
+};
+
+uint64_t
+round_code_bits(uint64_t length, uint64_t bits) {
+    return ((length + 7) / 8) * 8 * bits;
+}
+
+void
+normalize_for_cosine(const float* source, float* destination, uint64_t dim) {
```
Comment (LHT129) [inline src/quantization/saq_quantization/saq_quantizer.cpp:52]: **[suggestion]** `normalize_for_cosine` only handles the `isfinite(norm) and is_approx_zero(norm)` case (zero/near-zero vectors → fill with 0). When `norm` is non-finite (NaN, Inf from `FP32ComputeIP` overflow), `std::isfinite(norm)` is false and the early-return is skipped, falling through to `Normalize` which may produce NaN outputs. Consider also handling the non-finite case explicitly, e.g. `if (not std::isfinite(norm) or is_approx_zero(norm))`.

File: src/quantization/saq_quantization/saq_quantizer.cpp (hunk 2 above)
Comment (Lss-lmj) [reply src/quantization/saq_quantization/saq_quantizer.cpp]: Thanks. `ComputeImpl` cannot return a boolean status because its API result is the distance (`float`); returning `false` would be converted to `0.0F` and silently report a valid zero distance. I kept the exception for invalid state, clarified this contract and corrected the error text in 58d628d, and added untrained/null-path coverage in 2f3320e.

File: src/quantization/saq_quantization/saq_quantizer.cpp (hunk 3 above)
Comment (Lss-lmj) [reply src/quantization/saq_quantization/saq_quantizer.cpp:52]: Thanks. Addressed in 58d628d: non-finite and near-zero cosine norms now both map to the zero-vector path. Regression coverage in 2f3320e covers NaN, infinity, and squared-norm overflow for both encoded-code and query-distance paths.

**Review by LHT129 — COMMENTED** [review]
## Code Review for PR #2792

This is a follow-up review after the author addressed the three issues raised by @LHT129 in the previous review round. All three issues have been properly resolved in commit `478b7b41`:

- `normalize_for_cosine` now handles non-finite norms (NaN/Inf) by mapping them to the zero-vector path.
- `ComputeImpl` now has a clear comment explaining why it throws instead of returning `false`.
- `ComputeDistWithThresholdImpl` now has a `constexpr bool enable_threshold` with a comment documenting the L2-only design.

### Additional observations:

**1. `ComputeImpl` cosine norm computation (line ~778-785)**

`ComputeImpl` computes `segment_norm1` and `segment_norm2` from `adjusted_scale` and the popcount-derived code statistics. However, `ComputeProjectedDistance` for cosine reads the stored `reconstructed_norm` from the second metadata float (`codes + segment.metadata_offset + sizeof(float)`). The header comment states: "cosine stores the reconstructed norm so query-to-code and code-to-code distances have identical semantics." While these two approaches should be mathematically equivalent, using the stored `reconstructed_norm` in `ComputeImpl` would guarantee bit-identical semantics and avoid any floating-point drift between the two code paths. Consider reading the stored norm from metadata in `ComputeImpl` for cosine as well.

**2. `adjusted_scale` zero fallback in `EncodeSegment` (line ~620)**

When `inner_product <= epsilon` after CAQ adjustment, `adjusted_scale` is set to `0.0F`, which zeroes out the entire segment. This is a valid degenerate case, but it means the segment contributes no information. Consider logging a warning when this occurs, as it may indicate problematic training data or parameter choices (e.g., too few bits for a high-variance segment).

**3. PCA eigenvector extraction fix — log message (pca_transformer.cpp line ~160)**

The error log message says `"Error in sgeqrf"` but the actual LAPACK call is `Ssyev`. This is a copy-paste artifact from the random orthogonal transformer code.

### Summary

The PR is in excellent shape. The three previously flagged issues are resolved. The three observations above are minor — none are correctness bugs, and all have existing test coverage. The implementation is well-architected, thoroughly tested (98.8% line coverage in the SAQ directory), and follows project conventions consistently.

File: src/quantization/saq_quantization/saq_quantizer.cpp [hunk 4]
```diff
@@ -0,0 +1,1117 @@
...
+                                      code_scale * code_offset * (code_sum1 + code_sum2) +
+                                      code_offset * code_offset * segment.length);
+        const double segment_norm1 = static_cast<double>(adjusted_scale1) * adjusted_scale1 *
+                                     (code_scale * code_scale * code_square_sum1 +
+                                      2.0 * code_scale * code_offset * code_sum1 +
+                                      code_offset * code_offset * segment.length);
+        const double segment_norm2 = static_cast<double>(adjusted_scale2) * adjusted_scale2 *
+                                     (code_scale * code_scale * code_square_sum2 +
+                                      2.0 * code_scale * code_offset * code_sum2 +
+                                      code_offset * code_offset * segment.length);
```
Comment (LHT129) [inline src/quantization/saq_quantization/saq_quantizer.cpp]: `ComputeImpl` computes `segment_norm1` and `segment_norm2` from `adjusted_scale` and popcount-derived code statistics. However, `ComputeProjectedDistance` for cosine reads the stored `reconstructed_norm` from the second metadata float. The header comment states cosine stores the reconstructed norm "so query-to-code and code-to-code distances have identical semantics." While mathematically equivalent, using the stored `reconstructed_norm` here would guarantee bit-identical semantics and avoid any floating-point drift between the two code paths.

File: src/quantization/saq_quantization/saq_quantizer.cpp [hunk 5]
```diff
@@ -0,0 +1,1117 @@
...
+        }
+        if (adjusted == 0) {
+            break;
+        }
+        inner_product = 0.0;
+        quantized_norm = 0.0;
+        for (uint64_t d = 0; d < segment.length; ++d) {
+            const double quantized = (scalar_codes[d] + 0.5) * delta - max_abs;
+            inner_product += data[d] * quantized;
+            quantized_norm += quantized * quantized;
```
Comment (LHT129) [inline src/quantization/saq_quantization/saq_quantizer.cpp:620]: When `inner_product <= epsilon` after CAQ adjustment, `adjusted_scale` is set to `0.0F`, which zeroes out the entire segment. This is a valid degenerate case, but the segment contributes no information. Consider logging a warning when this occurs, as it may indicate problematic training data or parameter choices (e.g., too few bits for a high-variance segment).

File: src/impl/transform/pca_transformer.cpp [hunk 6]
```diff
@@ -140,21 +142,29 @@ void
...
+        const auto chunk = static_cast<int32_t>(std::min(count - processed, max_chunk));
+        const float beta = processed == 0 ? 0.0F : 1.0F;
+        BlasFunction::Sgemm(BlasFunction::RowMajor,
+                            BlasFunction::Trans,
+                            BlasFunction::NoTrans,
+                            dim,
+                            dim,
+                            chunk,
+                            1.0F / static_cast<float>(count - 1),
+                            centralized_data + processed * input_dim_,
```
Comment (LHT129) [inline src/impl/transform/pca_transformer.cpp:160]: The error log message says `"Error in sgeqrf"` but the actual LAPACK call is `Ssyev`. This appears to be a copy-paste artifact from the random orthogonal transformer code (`random_orthogonal_transformer.cpp`).

File: src/quantization/saq_quantization/saq_quantizer.cpp (hunk 4 above)
Comment (Lss-lmj) [reply src/quantization/saq_quantization/saq_quantizer.cpp]: Thanks. Addressed in 58d628d. `ComputeImpl` now reads the stored reconstructed norm for both cosine codes, matching the metadata used by query-to-code distance, and avoids the no-longer-needed squared-code accumulation for non-L2 metrics. Regression coverage in 2f3320e scales the stored norm and verifies the expected cosine normalization.

File: src/quantization/saq_quantization/saq_quantizer.cpp (hunk 5 above)
Comment (Lss-lmj) [reply src/quantization/saq_quantization/saq_quantizer.cpp:620]: Thanks. I kept the zero-scale fallback and documented it as the canonical representation for zero or numerically degenerate segments in 58d628d. I did not add a per-segment warning because `EncodeSegment` runs for every vector segment and zero vectors are valid inputs; on million-vector builds that could flood logs and perturb performance. Existing zero/non-finite-vector coverage in 2f3320e verifies finite behavior. If observability is needed later, an aggregated counter would be safer than hot-path logging.

File: src/impl/transform/pca_transformer.cpp (hunk 6 above)
Comment (Lss-lmj) [reply src/impl/transform/pca_transformer.cpp:160]: Thanks. Fixed the copy-paste message to `Error in ssyev` in 58d628d and verified the PCA test suite.

**Review by LHT129 — COMMENTED** [review]
## Code Review Summary

This is a well-structured, high-quality PR that introduces SAQ (Segmented Code Adjustment Quantization) as a new opt-in base quantizer. The implementation is thorough with comprehensive test coverage (25+ test cases).

### What was reviewed
- **Core quantization**: `saq_quantizer.h/.cpp` — template-based SAQ quantizer with PCA rotation, variance-aware bit allocation via dynamic programming, CAQ code adjustment, and bit-plane packed storage
- **Transform fixes**: `pca_transformer.cpp` — fixed LAPACKE column-major eigenvector extraction and corrected error message from "sgeqrf" to "ssyev"
- **Random rotation**: `random_orthogonal_transformer.cpp` — added deterministic seed support and malformed-data validation
- **Integration**: parameter mapping across HGraph/IVF/Pyramid, flatten/bucket datacell support, serialization round-trip tests

### Key observations
1. The dynamic programming approach in `BuildDynamicPlan` for segment boundary and bit-width optimization is well-designed with proper tie-breaking (preferring fewer segments when error is comparable)
2. The `adjusted_scale = 0` fallback for numerically degenerate segments is a reasonable design choice, documented with a clear comment
3. The PCA eigenvector extraction fix correctly accounts for LAPACKE column-major storage semantics, and the new non-diagonal covariance test validates this
4. Bit-plane packed code storage with popcount-based distance computation avoids expensive inverse transforms
5. All public API failure paths follow their documented return contracts

### Test coverage
Extensive tests cover: dynamic/fixed segmentation, CAQ adjustment convergence, all three metric types (L2/IP/Cosine), serialization round-trips, random rotation reproducibility, malformed rotation rejection, batch-4 SIMD consistency, zero-norm cosine handling, progressive L2 threshold early termination, and bit-width packing for all supported widths (1-13 bits).

No blocking issues found.

File: src/algorithm/hgraph/hgraph_search.cpp [hunk 7]
```diff
@@ -847,4 +847,13 @@ HGraph::SearchWithRequest(const SearchRequest& request) const {
     return std::move(dataset_results);
```
Comment (LHT129) [inline src/algorithm/hgraph/hgraph_search.cpp:847]: [note] The `hgraph_search.cpp` change adds an explicit template instantiation for `brute_force_search<InnerSearchMode::KNN_SEARCH>`. This is needed because the template is defined in the header but called from the `.cpp` file with `std::optional<float>` threshold parameter.

File: src/quantization/saq_quantization/saq_quantizer.cpp [hunk 8]
```diff
@@ -0,0 +1,1130 @@
+// Copyright 2024-present the vsag project
```
Comment (LHT129) [inline src/quantization/saq_quantization/saq_quantizer.cpp:1]: [note] The `ComputeProjectedDistance` function signature takes `enable_threshold` as a parameter, but `ComputeDistWithThresholdImpl` uses `constexpr bool enable_threshold = metric == MetricType::METRIC_TYPE_L2SQR` to make the decision at compile time. The function-level `enable_threshold` parameter is always called with the same compile-time constant, so the runtime branch is dead code for non-L2 metrics. This is a clean design — just noting for future readers that the parameter exists for API symmetry with `ComputeProjectedDistance`.

**Review by LHT129 — COMMENTED** [review]
## Code Review for PR #2792

This is a follow-up review after the author addressed all issues raised in previous review rounds. All previously flagged issues have been properly resolved in the latest commit `0db6b2dc`:

- `normalize_for_cosine` now handles non-finite norms (NaN/Inf) by mapping them to the zero-vector path
- `ComputeImpl` now has a clear comment explaining why it throws instead of returning `false`
- `ComputeDistWithThresholdImpl` now uses `constexpr bool enable_threshold` with a comment documenting the L2-only design
- PCA transformer error message corrected from `sgeqrf` to `ssyev`
- `ComputeImpl` for cosine now reads the stored reconstructed norm from metadata for bit-identical semantics with query-to-code distance

### Additional observations

1. **Explicit template instantiation in hgraph_search.cpp**: The added `template DistHeapPtr HGraph::brute_force_search<InnerSearchMode::KNN_SEARCH>(...)` is the correct C++ pattern — the template is called from the `.cpp` file with `std::optional<float>` threshold, and without this explicit instantiation the linker would fail.

2. **`RECORD_OVERHEAD_BITS` constant**: Defined as `3 * sizeof(float) * 8 = 96` bits in the header. The per-segment metadata is 2 floats (scale + norm) = 64 bits, leaving 32 bits for record-level overhead. This is correctly accounted for in the budget calculations.

### Summary

The PR is in excellent shape. All previously flagged issues are resolved. The implementation is well-architected, thoroughly tested (25+ test cases covering dynamic/fixed segmentation, all three metrics, serialization round-trips, SIMD batch consistency, and edge cases), and follows project conventions consistently. No blocking issues found.

File: src/algorithm/hgraph/hgraph_search.cpp (hunk 7 above)
Comment (Lss-lmj) [reply src/algorithm/hgraph/hgraph_search.cpp:1298]: Thanks for the note. After rebasing onto the latest `main`, upstream now provides the KNN and RANGE explicit instantiations next to the template definition, so I removed the obsolete duplicate at the end of `hgraph_search.cpp`. The rebased build and focused SAQ tests pass.

**Comment by Lss-lmj** [comment]
Hi @wxyucs, this branch has been rebased onto the latest main and the conflicts have been resolved. Focused local validation passed:

- make fmt
- SAQ quantizer and parameter tests: 17 cases / 273 assertions
- SAQ datacell tests: 2 cases / 159 assertions
- HGraph, IVF, and Pyramid SAQ integration tests: 4 cases / 69 assertions
- PCA and random orthogonal transform tests: 5 cases / 21,100,412 assertions

Could you please add the kind/feature label and approve the GitHub Actions workflows for this fork PR when convenient? The version/1.1 label is already present.

Thank you.
